ostd/src/mm/page/meta.rs

```diff
@@ -92,196 +70,96 @@ pub(in crate::mm) struct MetaSlot {
     /// It is placed at the beginning of a slot because:
     ///  - the implementation can simply cast a `*const MetaSlot`
     ///    to a `*const PageMeta` for manipulation;
-    ///  - the subsequent fields can utilize the end padding of the
-    ///    the inner union to save space.
-    _inner: MetaSlotInner,
-    /// To store [`PageUsage`].
-    pub(super) usage: AtomicU8,
+    ///  - if the metadata need special alignment, we can provide
+    ///    at most `PAGE_METADATA_ALIGN` bytes of alignment;
+    ///  - the subsequent fields can utilize the padding of the
+    ///    reference count to save space.
+    storage: UnsafeCell<[u8; PAGE_METADATA_MAX_SIZE]>,
+    /// The reference count of the page.
     pub(super) ref_count: AtomicU32,
+    /// The virtual table that indicates the type of the metadata.
+    pub(super) vtable_ptr: UnsafeCell<PageMetaVtablePtr>,
 }
 
-pub(super) union MetaSlotInner {
-    _frame: ManuallyDrop<FrameMeta>,
-    _pt: ManuallyDrop<PageTablePageMeta>,
-}
+type PageMetaVtablePtr = core::ptr::DynMetadata<dyn PageMeta>;
 
-// Currently the sizes of the `MetaSlotInner` union variants are no larger
-// than 8 bytes and aligned to 8 bytes. So the size of `MetaSlot` is 16 bytes.
-//
 // Note that the size of `MetaSlot` should be a multiple of 8 bytes to prevent
 // cross-page accesses.
```

Comment:
Not sure why it prevents cross-page accesses? Perhaps you mean `PAGE_SIZE % META_SLOT_SIZE == 0`?